File: packages/domain-floorplan/src/opening.test.ts
import { describe, expect, it } from "vitest";
import { DomainError } from "./domain-error.js";
import { LengthMm } from "./length-mm.js";
import { Opening } from "./opening.js";
import { PointMm } from "./point-mm.js";
import { Wall } from "./wall.js";

function hostWall() {
  return Wall.create({
    id: "wall_host",
    start: PointMm.of(0, 0),
    end: PointMm.of(4000, 0),
    thickness: LengthMm.of(200),
    height: LengthMm.of(2800),
    kind: "exterior",
  });
}

describe("Opening", () => {
  it("places a rectangular door on a host wall", () => {
    const opening = Opening.create({
      id: "op_door",
      kind: "door",
      width: LengthMm.of(900),
      height: LengthMm.of(2100),
      offsetAlongWall: LengthMm.of(500),
      sillHeight: LengthMm.of(0),
    });
    const wall = hostWall().hostOpening(opening);
    expect(wall.openings).toHaveLength(1);
    expect(wall.openings[0]?.id).toBe("op_door");
  });

  it("rejects an opening that extends past the wall length", () => {
    const opening = Opening.create({
      id: "op_wide",
      kind: "door",
      width: LengthMm.of(2000),
      height: LengthMm.of(2100),
      offsetAlongWall: LengthMm.of(2500),
      sillHeight: LengthMm.of(0),
    });
    expect(() => hostWall().hostOpening(opening)).toThrow(DomainError);
  });

  it("rejects an opening taller than the host wall", () => {
    const opening = Opening.create({
      id: "op_tall",
      kind: "window",
      width: LengthMm.of(900),
      height: LengthMm.of(3000),
      offsetAlongWall: LengthMm.of(200),
      sillHeight: LengthMm.of(900),
    });
    expect(() => hostWall().hostOpening(opening)).toThrow(DomainError);
  });

  it("rejects non-positive opening width", () => {
    expect(() =>
      Opening.create({
        id: "op_zero",
        kind: "door",
        width: LengthMm.of(0),
        height: LengthMm.of(2100),
        offsetAlongWall: LengthMm.of(0),
        sillHeight: LengthMm.of(0),
      }),
    ).toThrow(DomainError);
  });
});

File: core/tests/opening_test.cpp
#include <gtest/gtest.h>

#include "toporoom/domain/error.hpp"
#include "toporoom/domain/length_mm.hpp"
#include "toporoom/domain/opening.hpp"
#include "toporoom/domain/point_mm.hpp"
#include "toporoom/domain/wall.hpp"

using toporoom::domain::DomainError;
using toporoom::domain::LengthMm;
using toporoom::domain::Opening;
using toporoom::domain::OpeningKind;
using toporoom::domain::OpeningProps;
using toporoom::domain::PointMm;
using toporoom::domain::Wall;
using toporoom::domain::WallKind;
using toporoom::domain::WallProps;

namespace {

Wall host_wall() {
  WallProps props;
  props.id = "wall_host";
  props.start = PointMm::of(0, 0);
  props.end = PointMm::of(4000, 0);
  props.thickness = LengthMm::of(200);
  props.height = LengthMm::of(2800);
  props.kind = WallKind::Exterior;
  return Wall::create(props);
}

OpeningProps door_props(const char* id, double width, double height, double offset,
                        double sill) {
  OpeningProps props;
  props.id = id;
  props.kind = OpeningKind::Door;
  props.width = LengthMm::of(width);
  props.height = LengthMm::of(height);
  props.offset_along_wall = LengthMm::of(offset);
  props.sill_height = LengthMm::of(sill);
  return props;
}

}  // namespace

TEST(Opening, PlacesDoorOnHostWall) {
  const Opening opening = Opening::create(door_props("op_door", 900, 2100, 500, 0));
  const Wall wall = host_wall().host_opening(opening);
  ASSERT_EQ(wall.openings().size(), 1u);
  EXPECT_EQ(wall.openings()[0].id(), "op_door");
}

TEST(Opening, RejectsOpeningPastWallLength) {
  const Opening opening = Opening::create(door_props("op_wide", 2000, 2100, 2500, 0));
  EXPECT_THROW(host_wall().host_opening(opening), DomainError);
}

TEST(Opening, RejectsOpeningTallerThanWall) {
  OpeningProps props = door_props("op_tall", 900, 3000, 200, 900);
  props.kind = OpeningKind::Window;
  const Opening opening = Opening::create(props);
  EXPECT_THROW(host_wall().host_opening(opening), DomainError);
}

TEST(Opening, DistinguishesArchwayFromDoor) {
  OpeningProps arch = door_props("op_arch", 1200, 2100, 200, 0);
  arch.kind = OpeningKind::Archway;
  const Opening opening = Opening::create(arch);
  EXPECT_EQ(opening.kind(), OpeningKind::Archway);
  EXPECT_NE(opening.kind(), OpeningKind::Door);
  const Wall wall = host_wall().host_opening(opening);
  EXPECT_EQ(wall.openings()[0].kind(), OpeningKind::Archway);
}
